pydantic/schema.py

```diff
@@ -0,0 +1,392 @@
+from datetime import date, datetime, time, timedelta
+from decimal import Decimal
+from enum import Enum
+from pathlib import Path
+from typing import Any, Dict, Sequence, Set, Tuple, Type
+from uuid import UUID
+
+from . import main
+from .fields import Field, Shape
+from .json import pydantic_encoder
+from .types import DSN, UUID1, UUID3, UUID4, UUID5, DirectoryPath, EmailStr, FilePath, Json, NameEmail, UrlStr
+from .utils import clean_docstring
+
+
+def get_flat_models_from_model(model: Type['main.BaseModel']) -> Set[Type['main.BaseModel']]:
+    flat_models: Set[Type[main.BaseModel]] = set()
+    assert issubclass(model, main.BaseModel)
+    flat_models.add(model)
+    for field in model.__fields__.values():
+        flat_models = flat_models | get_flat_models_from_field(field)
+    return flat_models
+
+
+def get_flat_models_from_field(field: Field) -> Set[Type['main.BaseModel']]:
+    flat_models: Set[Type[main.BaseModel]] = set()
+    if field.sub_fields:
+        flat_models = flat_models | get_flat_models_from_sub_fields(field.sub_fields)
+    elif isinstance(field.type_, type) and issubclass(field.type_, main.BaseModel):
+        flat_models = flat_models | get_flat_models_from_model(field.type_)
+    return flat_models
+
+
+def get_flat_models_from_sub_fields(fields) -> Set[Type['main.BaseModel']]:
+    flat_models: Set[Type[main.BaseModel]] = set()
+    for field in fields:
+        flat_models = flat_models | get_flat_models_from_field(field)
+    return flat_models
+
+
+def get_flat_models_from_models(models: Sequence[Type['main.BaseModel']]) -> Set[Type['main.BaseModel']]:
+    flat_models: Set[Type[main.BaseModel]] = set()
+    for model in models:
+        flat_models = flat_models | get_flat_models_from_model(model)
+    return flat_models
+
+
+def get_long_model_name(model: Type['main.BaseModel']):
+    assert issubclass(model, main.BaseModel)
+    prefix = model.__module__.replace('.', '__')
+    name = model.__name__
+    return f'{prefix}__{name}'
+
+
+def get_model_name_maps(
+    unique_models: Set[Type['main.BaseModel']]
+) -> Tuple[Dict[str, Type['main.BaseModel']], Dict[Type['main.BaseModel'], str]]:
```

Comment:
can we have some doc strings explaining what this does.

Reply:
Done